developer_guides/topology2/topology2.rst

```diff
@@ -0,0 +1,895 @@
+			# instance attribute
+			#
+			unique	"instance"
+		}
+	}
+
+Note that the pga object names are constructed with the class name "pga" followed by 2 attribute values, index
+and instance, ex: pga.1.1. Also note that both the attributes wil be given the type "integer" by default because
+they do not specify the type.
+
+1.1.3 Attribute default values
+******************************
+
+Optionally, class definitions can be extended to give default values for their attributes. Let's add a 
+"uuid" attribute of type string to the pga class above and give it a default value.
+
+.. code-block:: bash
+
+	Class.Widget."pga" {
+		#
+		# Pipeline ID for the pga widget object
+		#
+		DefineAttribute."index" {}
+
+		#
+		# pga object instance
+		#
+		DefineAttribute."instance" {}
+		
+		DefineAttribute."uuid" {
+			type "string"
+		}
+		
+		# attribute qualifiers
+		attributes {
+			#
+			# The PGA widget name would be constructed using the index and instance
+			# attributes. For ex: "pga.1.1" or "pga.10.2" etc.
+			#
+			!constructor [
+				"index"
+				"instance"
+			]
+			
+			#
+			# pga widget objects instantiated within the same alsaconf node must have unique
+			# instance attribute
+			#
+			unique	"instance"
+		}
+
+		# default attribute values		
+		uuid 			"7e:67:7e:b7:f4:5f:88:41:af:14:fb:a8:bd:bf:86:82"
+
+	}
+
+All pga objects will automatically be given the default uuid as specified above in the class definition.
+
+1.1.4 Advanced attribute qualifiers
+***********************************
+
+Apart from the basic attribute qualifiers, attributes in the class definition can be qualified
+using the following advanced keywords:
+
+* **Mandatory:** Attributes qualified as mandatory must be provided with a value in the object
+  instance. Objects with default values in the class definition need not be qualified as mandatory.
+  Also, note that attributes in the constructor array are mandatory by default as they are required
+  for building the object's name.
+
+* **Immutable:** Attribute values that are set in the class definition and cannot be modified in
+  the object instance.
+ 
+* **Deprecated:** Attributes that have been deprecated and should not be used in the object.
+
+Let's add some extra attributes and advanced qualifers into the pga class definition:
+
+.. code-block:: bash
+
+	Class.Widget."pga" {
+		# attribute definitions
+		DefineAttribute.instance {
+			type	integer
+		}
+		DefineAttribute.index {
+			type	integer
+		}
+		DefineAttribute."type" {
+			type	"string"
+		}
+		DefineAttribute."uuid" {
+			type	"string"
+		}
+		DefineAttribute."preload_count" {}
+		
+		# attribute qualifiers
+		attributes {
+			#
+			# The PGA widget name would be constructed using the index and instance
+			# attributes. For ex: "pga.1.1" or "pga.10.2" etc.
+			#
+			!constructor [
+				"index"
+				"instance"
+			]
+
+			#
+			# immutable attributes must be given default values and cannot be modified in the object instance
```

Comment:
use may (permission), shall (requirement) and should (recommendation).

do not use 'must' or 'would'.